[Backport release-25.05]  opencode: 0.0.46 -> 0.2.33

delafthi opened the pull request:
This PR cherry-picks all Opencode-related commits from the `master` branch to backport them to `release-25.05`. The update brings Opencode to the latest version to resolve confusion caused by the 25.05 release lacking the SST variant. For example, [x.com/bedesqui/status/1943047048867254365](https://x.com/bedesqui/status/1943047048867254365).

@GaetanLepage, what do you think? Should we use the most recent version for the backport, or would it be better to select a specific version instead?

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GaetanLepage committed:
opencode: 0.0.46 -> 0.0.52, switch to sst fork

(cherry picked from commit f917ffcbdf7ffc26340c572e2474be43dcd5fd8e)

GaetanLepage committed:
opencode: correct meta.homepage

(cherry picked from commit 70e5f1490916ca157181b4212519fed0c7c1d506)

delafthi committed:
maintainers: add delafthi

(cherry picked from commit f246c7f3e1e90eda55aa05c0b27f6810d8a59931)

delafthi committed:
opencode: 0.0.52 -> 0.1.169

BREAKING CHANGE: This version forked from the original opencode and was
partially rebuilt. Therefore, there are various breaking changes.

(cherry picked from commit f89a4b3cd805369480ca441e77fd234fad53b6bf)

delafthi committed:
opencode: 0.1.69 -> 0.1.176

(cherry picked from commit 4721ddd76f0ebd076f1e782ad3d22ec21f6bffe2)

delafthi committed:
opencode: 0.1.176 -> 0.1.181

- Replaced `stdenv` with `stdenvNoCC`
- Moved subpackages into main derivation to enable updating subpackages with `nix-update-script`
- Fixed version test, by providing a temporary `HOME` directory

(cherry picked from commit 1371762d80a2c4b71769d2fb32d06ac66ba568a4)

delafthi committed:
opencode: 0.1.181 -> 0.1.182

(cherry picked from commit 872fd0c102ca5f07a4bf140964f7f32ffdae6b20)

delafthi committed:
opencode: 0.1.182 -> 0.1.185

(cherry picked from commit 8ff5933ccc6b06b94afc97630b075c2400d9dfc7)

delafthi committed:
opencode: 0.1.182 -> 0.1.189

(cherry picked from commit e5d1fc72d26c1cd6ff66756119e3982cfea931c6)

delafthi committed:
opencode: 0.1.185 -> 0.1.194

(cherry picked from commit 4acc22caf616f5958042d2746921f39203277edc)

delafthi committed:
opencode: 0.1.194 -> 0.2.5

(cherry picked from commit 16426f94914eaa8509e5ae83cb97eb8451a4248f)

delafthi committed:
opencode: 0.2.5 -> 0.2.6

(cherry picked from commit d1a79fcd1714d6d32ede425e70dcbf1cc7cb132c)

delafthi committed:
opencode: 0.2.6 -> 0.2.13

(cherry picked from commit afa0c5dc82ec21b74169e1cf9d8787abcac976da)

delafthi committed:
opencode: 0.2.13 -> 0.2.20

(cherry picked from commit 7bc194686d26b9f8320c62d73ef8ee82457af8bc)

delafthi committed:
opencode: 0.2.20 -> 0.2.23

(cherry picked from commit 3e1661ea7ff1ad2b580296bc8ea3652fece8708a)

delafthi committed:
opencode: 0.2.23 -> 0.2.25

(cherry picked from commit a435da7902de33caef36e890f128e125f196bcf7)

delafthi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 424150`
Commit: `93720e316de84fae5067f42368e9dd4c5a2a949a`

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

GaetanLepage commented:
> @GaetanLepage, what do you think? Should we use the most recent version for the backport, or would it be better to select a specific version instead?

It's fine by me to backport the latest version. As long as we remain on `0.x`, it is not a major version update, so we can keep backporting.

GaetanLepage reviewed (approved):
Thank you!

drupol commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,40 +1,187 @@
+    outputHashMode = "recursive";
+  };
+
+  models-dev-data = fetchurl {
+    url = "https://models.dev/api.json";
+    sha256 = "sha256-igxQOC+Hz2FnXIW/S4Px9WhRuBhcIQIHO+7U8jHU1TQ=";
+  };
```
What if tomorrow that file is being updated?

drupol replied:
This is somehow fragile since they can decide to break the derivation by just changing the content of that file.

delafthi replied:
What alternatives do we have?

drupol replied:
Asking them to host the file on their repository? Hosting the file in NixOS/nixpkgs ?

drupol replied:
Not necessarily a package on its own, but retrieving the file from GitHub is already a good step forward.

delafthi replied:
Ah, I see the advantage now. Using a revision number or tag prevents arbitrary changes to `api.json` and allows us to lock the version reliably.

drupol replied:
Yes, you got the whole point.

delafthi replied:
In that case, I would prefer creating a new package. Including it here as another Bun module would add complexity and introduce four additional hashes to the project. Moreover, since `models.dev` is a separate project, it probably will be reused in other projects as well.

drupol replied:
Ok, do it within this PR, sort commits accordingly.

delafthi replied:
Created a draft PR for opencode (#425045) to keep track of this issue.

drupol commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,40 +1,187 @@
+
+  nativeBuildInputs = [ bun ];
+
+  patches = [ ./fix-models-macro.patch ];
```
A little comment explaining why this patch is needed would be nice.

delafthi replied:
resolved in #423443

drupol commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,40 +1,187 @@
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
```
You can remove this line, it is not needed.

delafthi replied:
resolved in #423443

drupol reviewed (commented):
Hello, I made a few comments, let me know if you need some clarifications.

delafthi committed:
opencode: remove unnecessary `mkdir` in `installPhase`

(cherry picked from commit dabbedd709d4e740a76b7e888a5a0bb6c983f3cd)

delafthi committed:
opencode: add comment for the necessity of `fix-model-macro.patch`

(cherry picked from commit 6f4fff59128c41282513415e2513b269afbc9bc5)

delafthi committed:
opencode: 0.2.25 -> 0.2.27

(cherry picked from commit ae30ba98d574b6618aca66457ec121388d439a58)

delafthi commented:
I’ve converted this PR to a draft until #423443 is merged. Once that’s done, I will cherry-pick the commits into this PR.

GaetanLepage commented:
> I’ve converted this PR to a draft until #423443 is merged. Once that’s done, I will cherry-pick the commits into this PR.

Yes, definitely.

delafthi committed:
opencode: 0.2.27 -> 0.2.33

Diff: https://github.com/sst/opencode/compare/refs/tags/v0.2.27...refs/tags/v0.2.33
(cherry picked from commit 1f0f25154225df0302adcd7b8110ad2c99e48adc)

delafthi commented:
#424373 has not been merged yet, but I believe we should not hold this PR back and proceed with it.

delafthi commented:
@drupol I suggest we merge this PR first and then proceed with backporting #424373, #425045, and #426050. Currently, I cannot backport #424373 because I will only be on the maintainers list after this PR is merged.
